exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -216,46 +220,60 @@ func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteReq
+		data, err := proto.Marshal(writeReq)
+		if err != nil {
+			return backoff.Permanent(consumererror.NewPermanent(err))
+		}
+		buf := make([]byte, len(data), cap(data))
+		compressedData := snappy.Encode(buf, data)
 
-	// Create the HTTP POST request to send to the endpoint
-	req, err := http.NewRequestWithContext(ctx, "POST", prwe.endpointURL.String(), bytes.NewReader(compressedData))
-	if err != nil {
-		return consumererror.NewPermanent(err)
-	}
+		// Create the HTTP POST request to send to the endpoint
+		req, err := http.NewRequestWithContext(ctx, "POST", prwe.endpointURL.String(), bytes.NewReader(compressedData))
+		if err != nil {
+			return backoff.Permanent(consumererror.NewPermanent(err))
+		}
 
-	// Add necessary headers specified by:
-	// https://cortexmetrics.io/docs/apis/#remote-api
-	req.Header.Add("Content-Encoding", "snappy")
-	req.Header.Set("Content-Type", "application/x-protobuf")
-	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
-	req.Header.Set("User-Agent", prwe.userAgentHeader)
+		// Add necessary headers specified by:
+		// https://cortexmetrics.io/docs/apis/#remote-api
+		req.Header.Add("Content-Encoding", "snappy")
+		req.Header.Set("Content-Type", "application/x-protobuf")
+		req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+		req.Header.Set("User-Agent", prwe.userAgentHeader)
 
-	resp, err := prwe.client.Do(req)
-	if err != nil {
-		return consumererror.NewPermanent(err)
-	}
-	defer resp.Body.Close()
+		resp, err := prwe.client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		// 2xx status code is considered a success
+		// 5xx errors are recoverable and the exporter should retry
+		// Reference for different behavior according to status code:
+		// https://github.com/prometheus/prometheus/pull/2552/files#diff-ae8db9d16d8057358e49d694522e7186
+		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			return nil
+		}
 
-	// 2xx status code is considered a success
-	// 5xx errors are recoverable and the exporter should retry
-	// Reference for different behavior according to status code:
-	// https://github.com/prometheus/prometheus/pull/2552/files#diff-ae8db9d16d8057358e49d694522e7186
-	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-		return nil
-	}
-	body, err := io.ReadAll(io.LimitReader(resp.Body, 256))
-	rerr := fmt.Errorf("remote write returned HTTP status %v; err = %w: %s", resp.Status, err, body)
-	if resp.StatusCode >= 500 && resp.StatusCode < 600 {
-		return rerr
+		body, err := io.ReadAll(io.LimitReader(resp.Body, 256))
+		rerr := fmt.Errorf("remote write returned HTTP status %v; err = %w: %s", resp.Status, err, body)
+		if resp.StatusCode >= 500 && resp.StatusCode < 600 {
+			return rerr
+		}
+		return backoff.Permanent(consumererror.NewPermanent(rerr))
 	}
-	return consumererror.NewPermanent(rerr)
+
+	// Use the BackOff instance to retry the operation with exponential backoff.
+	return backoff.Retry(retryFunc, backOff)
```

Comment:
don't you need to wrap the error from backoff.Retry into a `consumererror.NewPermanent` in case there is an error?

Reply:
I Think it is not necessary as per looking this [function](https://github.com/cenkalti/backoff/blob/v4.2.1/retry.go#L139-L144) it would just return what is given. we are using it as `backoff.Permanent(consumererror.NewPermanent(err))` it should work as expected. I had similar doubt and added a [check](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-519368654a873cc5b7d84aeaa1a8e01685710dd1893dd0df40cf0cee676ee5e5R1049) to see if it is return permanent error.
On the other hand we also append to [NewPermanent ](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38L206)Error here in all cases

Reply:
But I think it's still better to wrap in execute function itself. committed the changes.

Reply:
you were right @vasireddy99. I stand corrected. I was mislead by the previous version[ here.](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38L258)